packages/block-editor/src/components/use-block-sync/block-manager.js

```diff
@@ -0,0 +1,115 @@
+			// which was passed originally in the opposite direction.
+			if ( blockMap[ block.clientId ]?.[ direction ] === block ) {
+				return blockMap[ block.clientId ][ reversedDirection ];
+			}
+
+			const updatedBlock = {
+				...block,
+				innerBlocks: mapBlocks( block.innerBlocks, direction ),
+				clientId: mapBlockId( block.clientId, direction ),
+			};
+
+			// A reference to both the internal and external blocks are kept so
+			// that both can be compared by reference.
+			const blockRef = {
+				[ direction ]: block,
+				[ reversedDirection ]: updatedBlock,
+			};
+
+			// Make sure we can access that reference via either clientId.
+			blockMap[ block.clientId ] = blockRef;
+			blockMap[ updatedBlock.clientId ] = blockRef;
+
+			// Return the freshly updated block for use by the consumer.
+			return updatedBlock;
+		} );
+	};
+
+	const mapBlockId = ( blockClientId, direction ) => {
+		// Avoid running map code on non-string clientIds. There appear to be times
+		// where this function can be called with `blockClientId = {}`
+		if ( typeof blockClientId !== 'string' || ! shouldMapBlocks ) {
+			return blockClientId;
+		}
+
+		if ( ! clientIdMap[ direction ][ blockClientId ] ) {
+			const newId = uuid();
+			clientIdMap[ direction ][ blockClientId ] = newId;
+			clientIdMap[ flipDirection( direction ) ][ newId ] = blockClientId;
+		}
+		return clientIdMap[ direction ][ blockClientId ];
+	};
+
+	return {
+		updateOutgoingBlocks: ( blocks ) => mapBlocks( blocks, DIRECTION_OUT ),
+		updateIncomingBlocks: ( blocks ) => mapBlocks( blocks, DIRECTION_IN ),
+		getOutgoingBlockId: ( blockId ) => mapBlockId( blockId, DIRECTION_OUT ),
+		getIncomingBlockId: ( blockId ) => mapBlockId( blockId, DIRECTION_IN ),
+	};
+}
+
+const _entityListenerReference = {};
+function assocateWithEntity( entityId ) {
+	if ( ! entityId ) {
+		return false;
+	}
+
+	const hasExistingListener = !! _entityListenerReference[ entityId ];
+	if ( ! hasExistingListener ) {
+		_entityListenerReference[ entityId ] = true;
+	}
+	return hasExistingListener;
+}
+
+export default function useBlockManager( entityId ) {
+	// TODO: useState vs useRef here? Unsure if a change in the manager should
+	// re-run useBlockSync, or if pointing at the ref works fine.
+	const [ manager, updateManager ] = useState(
+		createBlockClientIdManager( assocateWithEntity( entityId ) )
+	);
+
+	const entityIdRef = useRef( entityId );
+
+	// This side effect creates a new block client ID manager if and only if the
+	// component starts using blocks from a new entity. Since consumers require
+	// the manager before this effect would be run, we initalize the manager
+	// above and then avoid initalizing a new one the first time the effect runs.
+	useEffect( () => {
+		if ( entityIdRef.current !== entityId ) {
+			updateManager(
+				createBlockClientIdManager( assocateWithEntity( entityId ) )
+			);
+			entityIdRef.current = entityId;
+		}
+	}, [ entityId ] );
```

Comment:
I might not know what I'm talking about, but couldn't this whole thing just be replaced with:
```js
	const manager = useMemo(
		() => createBlockClientIdManager( assocateWithEntity( entityId ) ),
		[ entityId ]
	);
```